### ISSUE TITLE
OptionalUnit false positive for Nothing return type

### ISSUE THREAD

**Issue by Mygod** [detekt/detekt#3864]
## Expected Behavior

```
protected open fun onClickNull(): Unit = throw UnsupportedOperationException()
```

This method should resolve to return type `Nothing` instead of `Unit`.

## Observed Behavior

Triggered `OptionalUnit` warning.

## Your Environment

* Version of detekt used: 1.14.2
* Version of Gradle used (if applicable):
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Codacy
* Link to your project (if it's a public repository): https://github.com/Mygod/VPNHotspot/blob/440b22faa19246ca4ab86345ea95d4d1fbc8d10a/mobile/src/main/java/be/mygod/vpnhotspot/manage/TetherManager.kt#L99

**Comment by schalkms** [comment]
The rule flags this method due to the Unit type being explicitly defined.

**Comment by Mygod** [comment]
If it wasn't explicitly defined, the method would not resolve to Unit return type.

**Comment by cortinico** [comment]
> If it wasn't explicitly defined, the method would not resolve to Unit return type.

Just out of curiosity: is there a specific reason why you want to specify the return type explicitly (e.g. you have Kotlin explicit API mode enabled or other?).

**Comment by Mygod** [comment]
Did you read my issue description? Without explicitly specifying the `kotlin.Unit` return type,

# the method return type would resolve to `kotlin.Nothing`.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-nothing.html

**Comment by Mygod** [comment]
Furthermore, removing the "optional" Unit return type will make the code not compile. Just try it.

**Comment by cortinico** [comment]
> Did you read my issue description?

Please don't be rude when answering to comments and assume good intent.

We (the project maintainers) are doing this in our free time and it's not great to receive such answers on a public forum.

Given that @t-kameyama already pushed a fix (thanks for doing it) we will most likely merge it. In normal circumstances we would have closed this issue without offering further support.

**Comment by schalkms** [comment]
Thanks for outlining this issue. Fix will be in v1.18.

**Comment by Mygod** [comment]
Thanks for the fix. Sorry for being rude but it was frustrating trying to repeat myself for the third time.
